File: docs/snippets/ov_common.c
// Copyright (C) 2018-2021 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//
#include <openvino/c/openvino.h>

void infer_request_callback(void* args) {
    // Operations after infer
}
void inputs_v10(ov_infer_request_t* infer_request) {
    //! [ov_api_2_0:get_input_tensor_v10]
    ov_tensor_t* input_tensor1 = NULL;
    ov_tensor_t* input_tensor2 = NULL;
    void* data = NULL;

    {
    // Get input tensor by index
    ov_infer_request_get_input_tensor_by_index(infer_request, 0, &input_tensor1);
    // IR v10 works with converted precisions (i64 -> i32)
    ov_tensor_data(input_tensor1, &data);
    int32_t* data1 = (int32_t*)data;
    // Fill first data ...
    }

    {
    // Get input tensor by tensor name
    ov_infer_request_get_tensor(infer_request, "data2_t", &input_tensor2);
    // IR v10 works with converted precisions (i64 -> i32)
    ov_tensor_data(input_tensor2, &data);
    int32_t* data2 = (int32_t*)data;
    // Fill first data ...
    }

    ov_tensor_free(input_tensor1);
    ov_tensor_free(input_tensor2);
    //! [ov_api_2_0:get_input_tensor_v10]
}

void inputs_aligned(ov_infer_request_t* infer_request) {
    //! [ov_api_2_0:get_input_tensor_aligned]
    ov_tensor_t* input_tensor1 = NULL;
    ov_tensor_t* input_tensor2 = NULL;
    void* data = NULL;
    {
    // Get input tensor by index
    ov_infer_request_get_input_tensor_by_index(infer_request, 0, &input_tensor1);
    // Element types, names and layouts are aligned with framework
    ov_tensor_data(input_tensor1, &data);
    // Fill first data ...
    }

    {
    // Get input tensor by tensor name
    ov_infer_request_get_tensor(infer_request, "data2_t", &input_tensor2);
    // Element types, names and layouts are aligned with framework
    ov_tensor_data(input_tensor2, &data);
    // Fill first data ...
    }

    ov_tensor_free(input_tensor1);
    ov_tensor_free(input_tensor2);
    //! [ov_api_2_0:get_input_tensor_aligned]
}

void outputs_v10(ov_infer_request_t* infer_request) {
    //! [ov_api_2_0:get_output_tensor_v10]
    ov_tensor_t* output_tensor = NULL;
    void* data = NULL;

    // model has only one output
    ov_infer_request_get_output_tensor(infer_request, &output_tensor);
    // IR v10 works with converted precisions (i64 -> i32)
    ov_tensor_data(output_tensor, &data);
    int32_t* out_data = (int32_t*)data;
    // process output data
    
    ov_tensor_free(output_tensor);
    //! [ov_api_2_0:get_output_tensor_v10]
}

void outputs_aligned(ov_infer_request_t* infer_request) {
    //! [ov_api_2_0:get_output_tensor_aligned]
    ov_tensor_t* output_tensor = NULL;
    void* out_data = NULL;
    
    // model has only one output
    ov_infer_request_get_output_tensor(infer_request, &output_tensor);
    // Element types, names and layouts are aligned with framework
    ov_tensor_data(output_tensor, &out_data);
    // process output data
    
    ov_tensor_free(output_tensor);
    //! [ov_api_2_0:get_output_tensor_aligned]
}

int main() {
    //! [ov_api_2_0:create_core]
    ov_core_t* core = NULL;
    ov_core_create(&core);
    //! [ov_api_2_0:create_core]

    //! [ov_api_2_0:read_model]
    ov_model_t* model = NULL;
    ov_core_read_model(core, "model.xml", NULL, &model);
    //! [ov_api_2_0:read_model]

    //! [ov_api_2_0:compile_model]
    ov_compiled_model_t* compiled_model = NULL;
    ov_core_compile_model(core, model, "CPU", 0, &compiled_model);
    //! [ov_api_2_0:compile_model]

    //! [ov_api_2_0:create_infer_request]
    ov_infer_request_t* infer_request = NULL;
    ov_compiled_model_create_infer_request(compiled_model, &infer_request);
    //! [ov_api_2_0:create_infer_request]

    inputs_aligned(infer_request);

    //! [ov_api_2_0:inference]
    ov_infer_request_infer(infer_request);
    //! [ov_api_2_0:inference]

    //! [ov_api_2_0:start_async_and_wait]
    // NOTE: For demonstration purposes we are trying to set callback
    ov_callback_t callback;
    callback.callback_func = infer_request_callback;
    callback.args = infer_request;
    ov_infer_request_set_callback(infer_request, &callback);
    // Start inference without blocking current thread
    ov_infer_request_start_async(infer_request);
    // Wait for inference completion
    ov_infer_request_wait(infer_request);
    // Wait for 10 milisecond
    ov_infer_request_wait_for(infer_request, 10);
    //! [ov_api_2_0:start_async_and_wait]

    outputs_aligned(infer_request);

    //! [ov_api_2_0:load_old_extension]
    // For C API 2.0 "add_extension()" is not supported for now
    //! [ov_api_2_0:load_old_extension]
    ov_infer_request_free(infer_request);
    ov_compiled_model_free(compiled_model);
    ov_model_free(model);
    ov_core_free(core);
    return 0;
}

File: docs/snippets/ov_common.py
# Copyright (C) 2022 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
#


import numpy as np
from utils import get_image, get_path_to_extension_library, get_path_to_model

#! [ov_api_2_0:create_core]
import openvino as ov
core = ov.Core()
#! [ov_api_2_0:create_core]

model_path = get_path_to_model()

#! [ov_api_2_0:read_model]
model = core.read_model(model_path)
#! [ov_api_2_0:read_model]

#! [ov_api_2_0:compile_model]
compiled_model = core.compile_model(model, "CPU")
#! [ov_api_2_0:compile_model]

#! [ov_api_2_0:create_infer_request]
infer_request = compiled_model.create_infer_request()
#! [ov_api_2_0:create_infer_request]

#! [ov_api_2_0:get_input_tensor_aligned]
# Get input tensor by index
input_tensor1 = infer_request.get_input_tensor(0)
# Element types, names and layouts are aligned with framework
assert input_tensor1.data.dtype == np.int64
# Fill the first data ...

# Get input tensor by tensor name
input_tensor2 = infer_request.get_tensor("input")
assert input_tensor2.data.dtype == np.int64
# Fill the second data ...
#! [ov_api_2_0:get_input_tensor_aligned]

#! [ov_api_2_0:get_input_tensor_v10]
# Get input tensor by index
input_tensor1 = infer_request.get_input_tensor(0)
# IR v10 works with converted precisions (i64 -> i32)
assert input_tensor1.data.dtype == np.int32
# Fill the first data ...

# Get input tensor by tensor name
input_tensor2 = infer_request.get_tensor("input")
# IR v10 works with converted precisions (i64 -> i32)
assert input_tensor2.data.dtype == np.int32
# Fill the second data ..
#! [ov_api_2_0:get_input_tensor_v10]

#! [ov_api_2_0:inference]
results = infer_request.infer()
#! [ov_api_2_0:inference]

input_data = get_image()

def process_results(results, frame_id):
    pass

#! [ov_api_2_0:start_async_and_wait]
# Start async inference on a single infer request
infer_request.start_async()
# Wait for 1 milisecond
infer_request.wait_for(1)
# Wait for inference completion
infer_request.wait()

# Demonstrates async pipeline using AsyncInferQueue

results = []

def callback(request, frame_id):
    # Copy the data from output tensors to numpy array and process it
    results_copy = {output: data[:] for output, data in request.results.items()}
    results.append(process_results(results_copy, frame_id))

# Create AsyncInferQueue with 4 infer requests
infer_queue = ov.AsyncInferQueue(compiled_model, jobs=4)
# Set callback for each infer request in the queue
infer_queue.set_callback(callback)

total_frames = 100
for i in range(total_frames):
    # Wait for at least one available infer request and start asynchronous inference
    infer_queue.start_async(input_data, userdata=i)
# Wait for all requests to complete
infer_queue.wait_all()
#! [ov_api_2_0:start_async_and_wait]

#! [ov_api_2_0:get_output_tensor_aligned]
# Model has only one output
output_tensor = infer_request.get_output_tensor()
# Element types, names and layouts are aligned with framework
assert output_tensor.data.dtype == np.int64
# process output data ...
#! [ov_api_2_0:get_output_tensor_aligned]

#! [ov_api_2_0:get_output_tensor_v10]
# Model has only one output
output_tensor = infer_request.get_output_tensor()
# IR v10 works with converted precisions (i64 -> i32)
assert output_tensor.data.dtype == np.int32
# process output data ...
#! [ov_api_2_0:get_output_tensor_v10]

path_to_extension_library = get_path_to_extension_library()

#! [ov_api_2_0:load_old_extension]
core.add_extension(path_to_extension_library)
#! [ov_api_2_0:load_old_extension]
